SDWebImage/SDImageAPNGCoder.m

```diff
@@ -81,8 +81,9 @@ - (UIImage *)decodedImageWithData:(NSData *)data options:(nullable SDImageCoderO
         return nil;
     }
     CGFloat scale = 1;
-    if ([options valueForKey:SDImageCoderDecodeScaleFactor]) {
-        scale = [[options valueForKey:SDImageCoderDecodeScaleFactor] doubleValue];
+    NSNumber *scaleFactor = options[SDImageCoderDecodeScaleFactor];
```

Comment:
This change. Actually take no big benefit. All our key used for dictionary, are not suffix with `@`. So it does not trigger `-[NSObject valueForKey:]` at all. So both of this is OK...

Actually. These changes may only about coding style :)

Reply:
@dreampiggy Maybe you misleading my description, please see [`Apple Documents`](https://developer.apple.com/documentation/foundation/nsdictionary/1410210-valueforkey)
.

Reply:
```
Foundation`-[NSDictionary(NSKeyValueCoding) valueForKey:]:
->  0x108165510 <+0>:   push   rbp
    0x108165511 <+1>:   mov    rbp, rsp
    0x108165514 <+4>:   push   r15
    0x108165516 <+6>:   push   r14
    0x108165518 <+8>:   push   rbx
    0x108165519 <+9>:   sub    rsp, 0x18
    0x10816551d <+13>:  mov    r15, rdx
    0x108165520 <+16>:  mov    r14, rdi
    0x108165523 <+19>:  mov    rsi, qword ptr [rip + 0x384cd6] ; "length"
    0x10816552a <+26>:  mov    rdi, r15
    0x10816552d <+29>:  call   qword ptr [rip + 0x2d72fd] ; (void *)0x000000010876f980: objc_msgSend
    0x108165533 <+35>:  mov    rbx, rax
    0x108165536 <+38>:  test   rbx, rbx
    0x108165539 <+41>:  je     0x1081655c1               ; <+177>
    0x10816553f <+47>:  mov    rsi, qword ptr [rip + 0x384ff2] ; "characterAtIndex:"
    0x108165546 <+54>:  xor    edx, edx
    0x108165548 <+56>:  mov    rdi, r15
    0x10816554b <+59>:  call   qword ptr [rip + 0x2d72df] ; (void *)0x000000010876f980: objc_msgSend
    0x108165551 <+65>:  movzx  eax, ax
    0x108165554 <+68>:  cmp    eax, 0x40
    0x108165557 <+71>:  jne    0x1081655c1               ; <+177>
    0x108165559 <+73>:  dec    rbx
    0x10816555c <+76>:  mov    rsi, qword ptr [rip + 0x38520d] ; "_newSubstringWithRange:zone:"
    0x108165563 <+83>:  mov    edx, 0x1
    0x108165568 <+88>:  xor    r8d, r8d
    0x10816556b <+91>:  mov    rdi, r15
    0x10816556e <+94>:  mov    rcx, rbx
    0x108165571 <+97>:  call   qword ptr [rip + 0x2d72b9] ; (void *)0x000000010876f980: objc_msgSend
    0x108165577 <+103>: mov    rbx, rax
    0x10816557a <+106>: test   rbx, rbx
    0x10816557d <+109>: je     0x1081655c1               ; <+177>
    0x10816557f <+111>: lea    rdi, [rbp - 0x28]
    0x108165583 <+115>: mov    qword ptr [rdi], r14
    0x108165586 <+118>: mov    rax, qword ptr [rip + 0x38eafb] ; (void *)0x000000010938b288: NSDictionary
    0x10816558d <+125>: mov    qword ptr [rdi + 0x8], rax
    0x108165591 <+129>: mov    rsi, qword ptr [rip + 0x386628] ; "valueForKey:"
    0x108165598 <+136>: mov    rdx, rbx
    0x10816559b <+139>: call   0x1083a81dc               ; symbol stub for: objc_msgSendSuper2
    0x1081655a0 <+144>: mov    r14, rax
    0x1081655a3 <+147>: mov    rsi, qword ptr [rip + 0x3849f6] ; "release"
    0x1081655aa <+154>: mov    rdi, rbx
    0x1081655ad <+157>: call   qword ptr [rip + 0x2d727d] ; (void *)0x000000010876f980: objc_msgSend
    0x1081655b3 <+163>: mov    rax, r14
    0x1081655b6 <+166>: add    rsp, 0x18
    0x1081655ba <+170>: pop    rbx
    0x1081655bb <+171>: pop    r14
    0x1081655bd <+173>: pop    r15
    0x1081655bf <+175>: pop    rbp
    0x1081655c0 <+176>: ret    
    0x1081655c1 <+177>: mov    rsi, qword ptr [rip + 0x384b10] ; "objectForKey:"
    0x1081655c8 <+184>: mov    rdi, r14
    0x1081655cb <+187>: mov    rdx, r15
    0x1081655ce <+190>: add    rsp, 0x18
    0x1081655d2 <+194>: pop    rbx
    0x1081655d3 <+195>: pop    r14
    0x1081655d5 <+197>: pop    r15
    0x1081655d7 <+199>: pop    rbp
    0x1081655d8 <+200>: jmp    qword ptr [rip + 0x2d7252] ; (void *)0x000000010876f980: objc_msgSend

```

Reply:
@zhongwuzw I know...I means, there are not so many performance issue on this. It just check `NSString.characterAtIndex(0) == @"@"`, and call super instead...

And, Foundation framework is not private framework, the implementation can be found or disassembled using Hopper, don't need to ready ARM assemble code :)

```objectivec
/* @class NSDictionary */
-(void *)valueForKey:(void *)arg2 {
    r15 = arg2;
    r14 = self;
    rbx = [r15 length];
    if ((rbx != 0x0) && (([r15 characterAtIndex:0x0] & 0xffff) == 0x40)) {
            r12 = [r15 _newSubstringWithRange:0x1 zone:rbx - 0x1];
            if (r12 != 0x0) {
                    var_30 = r14;
                    rax = [[&var_30 super] valueForKey:r12];
                    rbx = rax;
                    [r12 release];
            }
            else {
                    rbx = [r14 objectForKey:r15];
            }
    }
    else {
            rbx = [r14 objectForKey:r15];
    }
    rax = rbx;
    return rax;
}
```

Anyway, we can change this without no issue...

Reply:
> Seems sometimes you're worrying too much about performance 😅

@dreampiggy 😂 , yeah, maybe, in my opinion, we all need try my best to write more modular, efficiency and steady code, and keep a good coding style. Because there are huge iOS developers to use, at the same time, to learn. 

Back to these changes, every option when we access, we call `valueForKey:` two times, it's not necessary; and in our situations, `valueForKey:` actually call `objectForKey:` internally. So let's try to do the better ways. 😄

Reply:
"early optimisations are the root of all evil" :))

In this particular case, I think `objectForKey` is better than `valueForKey` for the following reasons:
- code readability and semantics (since `objectForKey` can be easily used via the `[]` subscription operator)
- the definition `- (nullable ObjectType)objectForKey:(KeyType)aKey` shows that `objectForKey` is more strict in terms of types, especially is used on a dictionary with generics `NSDictionary<String,String>`. `- (nullable id)valueForKey:(NSString *)key` is part of `NSObject(NSKeyValueCoding)`, while `objectForKey` is defined on `NSDictionary`.
- indeed, a very small performance difference - not important here as we have very few calls.